blog/2024-02-02-migrating-machines-to-xstate-v5/index.mdx

```diff
@@ -0,0 +1,651 @@
+
+1. [Install XState v5 and dependencies](#1-install-xstate-v5-and-dependencies)
+2. [Move types to `setup()` and remove typegen](#2-move-types-to-setup-and-remove-typegen)
+3. [Convert action and guard strings to parameterized objects](#3-convert-action-and-guard-strings-to-parameterized-objects)
+4. [Include implementations or stubs in `setup()`](#4-include-implementations-or-stubs-in-setup)
+5. [Provide concrete implementations](#5-provide-concrete-implementations)
+
+## 1. Install XState v5 and dependencies
+
+The first step will be to [install XState v5](/docs/xstate#installation). You can also install the library-specific dependencies that are compatible with XState v5 if you’re using [React](/docs/xstate-react), [Vue](/docs/xstate-vue), or [Svelte](/docs/xstate-svelte).
+
+If you have many state machines and would like to incrementally migrate those to v5, you can install both v4 and v5 side-by-side by [following the migration steps listed here](/docs/migration#how-can-i-use-both-xstate-v4-and-v5). Once you’ve completed those steps, you’ll end up with both dependencies in your `package.json`. Here’s an example with libraries pegged to the latest version at the time of this writing but you can also use `npm:xstate@latest` if you prefer.
+
+:::tip
+
+Alternatively, you can also default `xstate` to v5 and use `xstate4` for v4.
+
+:::
+
+```json5
+// package.json after completing the steps in the migration guide
+{
+  dependencies: {
+    xstate: '4.38.2',
+    xstate5: 'npm:xstate@5.6.0',
+    '@xstate/react': '3.2.2',
+    '@xstate/react5': 'npm:@xstate/react@4.0.3',
+  },
+}
+```
+
+If using a dual installation (v5 alongside of v4), or even if only using v5, you will need to make sure you import the XState functions from the v5 packages in your code. At the top of your file, import functions like `setup`, `assign`, and other action creators from a `xstatev5` package.
+
+:::warning
+
+Calling `setup().createMachine()` from XState v5 but then calling `assign` from v4 within a v5 machine could result in type errors.
+
+:::
+
+## 2. Move types to `setup()` and remove typegen
+
+One of the first migration steps, based on a breaking change, is to remove any TypeScript types from the `schema` property of the machine config from v4. These types should now be [included under a new `types` property](/docs/typescript#specifying-types) in the object passed to the new `setup({})` function.
+
+<Tabs>
+<TabItem value="v5" label="XState v5">
+
+```ts
+import { setup } from 'xstate';
+
+const machine = setup({
+  types: {} as {
+    context: {
+      /* type context props */
+    };
+    events: {
+      /* type events */
+    };
+  },
+  /* implementations */
+  actions: {},
+  guards: {},
+  actors: {},
+}).createMachine({
+  /* machine config */
+});
+```
+
+</TabItem>
+
+<TabItem value="v4" label="XState v4">
+
+```ts
+// ❌ DEPRECATED
+import { createMachine } from 'xstate';
```

Comment:
It's not exactly deprecated - we still have this API and we didn't deprecate it. `setup` is preferred though since it automates converting implementations for actions (and more) to types.

Reply:
Great. Thanks for this clarification. I'll remove "deprecated" and add mention of the reasoning as well.